In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced

from sklearn.preprocessing import StandardScaler,OneHotEncoder

## Import Provisional CSV and Perform Basic Data Cleaning

In [4]:
# # Load the data
# file_name = "Resources/Diamonds_Prices2022.csv"
# df = pd.read_csv(file_name, index_col = "Unnamed: 0")

In [5]:
from config import db_password
import sqlalchemy as db
import psycopg2

db_string = f"postgresql://postgres:{db_password}@127.0.0.1:5432/Diamonds_Price_data"
engine = db.create_engine(db_string)
connection = engine.connect()
metadata = db.MetaData()

diamonds_prices2022_table = db.Table('diamonds_prices2022', metadata, autoload=True, autoload_with=engine)
#Equivalent to 'SELECT * FROM diamonds_prices2022_table'
query = db.select([diamonds_prices2022_table])
ResultProxy = connection.execute(query)
ResultSet = ResultProxy.fetchall()
df = pd.DataFrame(ResultSet)
df.columns = ResultSet[0].keys()
df = df.drop("id", axis=1)
df.columns
df

carat        cut color clarity depth table  price     x     y     z
0      0.23      Ideal     E     SI2  61.5    55    326  3.95  3.98  2.43
1      0.21    Premium     E     SI1  59.8    61    326  3.89  3.84  2.31
2      0.23       Good     E     VS1  56.9    65    327  4.05  4.07  2.31
3      0.29    Premium     I     VS2  62.4    58    334   4.2  4.23  2.63
4      0.31       Good     J     SI2  63.3    58    335  4.34  4.35  2.75
...     ...        ...   ...     ...   ...   ...    ...   ...   ...   ...
53938  0.86    Premium     H     SI2    61    58   2757  6.15  6.12  3.74
53939  0.75      Ideal     D     SI2  62.2    55   2757  5.83  5.87  3.64
53940  0.71    Premium     E     SI1  60.5    55   2756  5.79  5.74  3.49
53941  0.71    Premium     F     SI1  59.8    62   2756  5.74  5.73  3.43
53942   0.7  Very Good     E     VS2  60.5    59   2757  5.71  5.76  3.47

[53943 rows x 10 columns]

In [6]:
# EDA showed no missing values

# # Drop the null columns where all values are null
# df = df.dropna(axis='columns', how='all')

# # Drop the null rows
# df = df.dropna()

#### Address z == 0
Missing values in x, y, and z were assigned as 0's

In [7]:
# all x and y 0's also have z 0's

# find (x,y,z) 0's
df[df.z == 0]

carat        cut color clarity depth table  price     x     y  z
2207      1    Premium     G     SI2  59.1    59   3142  6.55  6.48  0
2314   1.01    Premium     H      I1  58.1    59   3167  6.66   6.6  0
4791    1.1    Premium     G     SI2    63    59   3696   6.5  6.47  0
5471   1.01    Premium     F     SI2  59.2    58   3837   6.5  6.47  0
10167   1.5       Good     G      I1    64    61   4731  7.15  7.04  0
11182  1.07      Ideal     F     SI2  61.6    56   4954     0  6.62  0
11963     1  Very Good     H     VS2  63.3    53   5139     0     0  0
13601  1.15      Ideal     G     VS2  59.2    56   5564  6.88  6.83  0
15951  1.14       Fair     G     VS1  57.5    67   6381     0     0  0
24394  2.18    Premium     H     SI2  59.4    61  12631  8.49  8.45  0
24520  1.56      Ideal     G     VS2  62.2    54  12800     0     0  0
26123  2.25    Premium     I     SI1  61.3    58  15397  8.52  8.42  0
26243   1.2    Premium     D    VVS1  62.1    59  15686     0     0  0
27112   2.2    Premium     H     SI1  61.2    59  17265  8.42  8.37  0
27429  2.25    Premium     H     SI2  62.8    59  18034     0     0  0
27503  2.02    Premium     H     VS2  62.7    53  18207  8.02  7.95  0
27739   2.8       Good     G     SI2  63.8    58  18788   8.9  8.85  0
49556  0.71       Good     F     SI2  64.1    60   2130     0     0  0
49557  0.71       Good     F     SI2  64.1    60   2130     0     0  0
51506  1.12    Premium     G      I1  60.4    59   2383  6.71  6.67  0

In [8]:
# z can not be recovered without x, 
# as round cut diamonds can be quite oval

df.drop(df[df.x == 0].index, inplace=True)

df[df.z == 0]

carat      cut color clarity depth table  price     x     y  z
2207      1  Premium     G     SI2  59.1    59   3142  6.55  6.48  0
2314   1.01  Premium     H      I1  58.1    59   3167  6.66   6.6  0
4791    1.1  Premium     G     SI2    63    59   3696   6.5  6.47  0
5471   1.01  Premium     F     SI2  59.2    58   3837   6.5  6.47  0
10167   1.5     Good     G      I1    64    61   4731  7.15  7.04  0
13601  1.15    Ideal     G     VS2  59.2    56   5564  6.88  6.83  0
24394  2.18  Premium     H     SI2  59.4    61  12631  8.49  8.45  0
26123  2.25  Premium     I     SI1  61.3    58  15397  8.52  8.42  0
27112   2.2  Premium     H     SI1  61.2    59  17265  8.42  8.37  0
27503  2.02  Premium     H     VS2  62.7    53  18207  8.02  7.95  0
27739   2.8     Good     G     SI2  63.8    58  18788   8.9  8.85  0
51506  1.12  Premium     G      I1  60.4    59   2383  6.71  6.67  0

In [9]:
# z can be recovered from x, y, and depth:
# z = depth / 100 * mean(x, y)

index_values = df[df.z == 0].index.values
print(index_values)

df.z = np.where(df.z == 0, df.depth / 100 * ((df.x + df.y) / 2), df.z)

# updated z values that were previously 0's
df.loc[index_values]

[ 2207  2314  4791  5471 10167 13601 24394 26123 27112 27503 27739 51506]


carat      cut color clarity depth table  price     x     y         z
2207      1  Premium     G     SI2  59.1    59   3142  6.55  6.48  3.850365
2314   1.01  Premium     H      I1  58.1    59   3167  6.66   6.6   3.85203
4791    1.1  Premium     G     SI2    63    59   3696   6.5  6.47   4.08555
5471   1.01  Premium     F     SI2  59.2    58   3837   6.5  6.47  3.839120
10167   1.5     Good     G      I1    64    61   4731  7.15  7.04   4.54080
13601  1.15    Ideal     G     VS2  59.2    56   5564  6.88  6.83  4.058160
24394  2.18  Premium     H     SI2  59.4    61  12631  8.49  8.45   5.03118
26123  2.25  Premium     I     SI1  61.3    58  15397  8.52  8.42   5.19211
27112   2.2  Premium     H     SI1  61.2    59  17265  8.42  8.37  5.137740
27503  2.02  Premium     H     VS2  62.7    53  18207  8.02  7.95  5.006595
27739   2.8     Good     G     SI2  63.8    58  18788   8.9  8.85  5.662250
51506  1.12  Premium     G      I1  60.4    59   2383  6.71  6.67   4.04076

#### Possible feature selection
- Combine colors I + J, as they contain relatively few values..
- Combine clarities IF + VVS1 and I1 + SI2, as IF and I1 have relatively few values, and IF has a similar clarity to VVS21, and I1 has a similar clarity to SI2 

In [10]:
# encode feature columns that are strings/objects
# use get_dummies

type_objs = ['cut', 'color', 
             'clarity'
            ]

df = pd.get_dummies(df, columns=type_objs)

print(df.shape)
df.head()

(53935, 27)


carat depth table  price     x     y     z  cut_Fair  cut_Good  cut_Ideal  \
0  0.23  61.5    55    326  3.95  3.98  2.43         0         0          1   
1  0.21  59.8    61    326  3.89  3.84  2.31         0         0          0   
2  0.23  56.9    65    327  4.05  4.07  2.31         0         1          0   
3  0.29  62.4    58    334   4.2  4.23  2.63         0         0          0   
4  0.31  63.3    58    335  4.34  4.35  2.75         0         1          0   

   ...  color_I  color_J  clarity_I1  clarity_IF  clarity_SI1  clarity_SI2  \
0  ...        0        0           0           0            0            1   
1  ...        0        0           0           0            1            0   
2  ...        0        0           0           0            0            0   
3  ...        1        0           0           0            0            0   
4  ...        0        1           0           0            0            1   

   clarity_VS1  clarity_VS2  clarity_VVS1  clarity_VVS2  
0            0            0             0             0  
1            0            0             0             0  
2            1            0             0             0  
3            0            1             0             0  
4            0            0             0             0  

[5 rows x 27 columns]

In [11]:
df.columns

Index(['carat', 'depth', 'table', 'price', 'x', 'y', 'z', 'cut_Fair',
       'cut_Good', 'cut_Ideal', 'cut_Premium', 'cut_Very Good', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_I1', 'clarity_IF', 'clarity_SI1', 'clarity_SI2', 'clarity_VS1',
       'clarity_VS2', 'clarity_VVS1', 'clarity_VVS2'],
      dtype='object')

In [12]:
# reorder columns by diamond trait quality
num_cols = list(df.columns[:7])
cut_cols = ['cut_Ideal', 'cut_Premium', 'cut_Very Good', 'cut_Good', 'cut_Fair']
color_cols = list(df.columns[12:19])
clarity_cols = ['clarity_IF', 
                'clarity_VVS1', 'clarity_VVS2', 
                'clarity_VS1', 'clarity_VS2', 
                'clarity_SI1', 'clarity_SI2', 
                'clarity_I1']
new_col_order = num_cols + cut_cols + color_cols + clarity_cols
df = df[new_col_order]
df.columns

Index(['carat', 'depth', 'table', 'price', 'x', 'y', 'z', 'cut_Ideal',
       'cut_Premium', 'cut_Very Good', 'cut_Good', 'cut_Fair', 'color_D',
       'color_E', 'color_F', 'color_G', 'color_H', 'color_I', 'color_J',
       'clarity_IF', 'clarity_VVS1', 'clarity_VVS2', 'clarity_VS1',
       'clarity_VS2', 'clarity_SI1', 'clarity_SI2', 'clarity_I1'],
      dtype='object')

# Split the Data into Training and Testing

In [13]:
# Create our features
X = df.copy()
X = X.drop('price', axis=1)

# Create our target
y = df["price"].values

In [14]:
X.describe()

cut_Ideal   cut_Premium  cut_Very Good      cut_Good      cut_Fair  \
count  53935.000000  53935.000000   53935.000000  53935.000000  53935.000000   
mean       0.399536      0.255697       0.224010      0.090924      0.029832   
std        0.489808      0.436256       0.416933      0.287504      0.170126   
min        0.000000      0.000000       0.000000      0.000000      0.000000   
25%        0.000000      0.000000       0.000000      0.000000      0.000000   
50%        0.000000      0.000000       0.000000      0.000000      0.000000   
75%        1.000000      1.000000       0.000000      0.000000      0.000000   
max        1.000000      1.000000       1.000000      1.000000      1.000000   

            color_D       color_E      color_F       color_G       color_H  \
count  53935.000000  53935.000000  53935.00000  53935.000000  53935.000000   
mean       0.125596      0.181682      0.17688      0.209326      0.153926   
std        0.331396      0.385586      0.38157      0.406831      0.360881   
min        0.000000      0.000000      0.00000      0.000000      0.000000   
25%        0.000000      0.000000      0.00000      0.000000      0.000000   
50%        0.000000      0.000000      0.00000      0.000000      0.000000   
75%        0.000000      0.000000      0.00000      0.000000      0.000000   
max        1.000000      1.000000      1.00000      1.000000      1.000000   

            color_I       color_J    clarity_IF  clarity_VVS1  clarity_VVS2  \
count  53935.000000  53935.000000  53935.000000  53935.000000  53935.000000   
mean       0.100528      0.052063      0.033188      0.067748      0.093928   
std        0.300706      0.222155      0.179129      0.251316      0.291731   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

        clarity_VS1   clarity_VS2   clarity_SI1   clarity_SI2    clarity_I1  
count  53935.000000  53935.000000  53935.000000  53935.000000  53935.000000  
mean       0.151479      0.227255      0.242273      0.170390      0.013739  
std        0.358518      0.419063      0.428463      0.375979      0.116406  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000      0.000000      0.000000  
75%        0.000000      0.000000      0.000000      0.000000      0.000000  
max        1.000000      1.000000      1.000000      1.000000      1.000000

In [15]:
# Check the balance of our target values
# Counter(y)

In [16]:
# Splitting into Train and Test sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size=0.25, # default
                                                    random_state=1)

In [17]:
# # Create a StandardScaler instances
# scaler = StandardScaler()

# # Fit the StandardScaler
# X_scaler = scaler.fit(X_train)

# # Scale the data
# X_train_scaled = X_scaler.transform(X_train)
# X_test_scaled = X_scaler.transform(X_test)

In [18]:
from sklearn.linear_model import LinearRegression

In [19]:
lm = LinearRegression()
lm.fit(X_train, y_train)

LinearRegression()

In [20]:
# R-squared values
print(f"Training data R-squared: {lm.score(X_train, y_train)}")
print(f"Testing data R-squared:  {lm.score(X_test, y_test)}")

Training data R-squared: 0.9196831285444621
Testing data R-squared:  0.9214680077548184


In [21]:
print(f"Training data intercept: {lm.intercept_}")
print(f"Training data coefficients: {lm.coef_}")

Training data intercept: 5790.202354305243
Training data coefficients: [11403.54429269   -61.97280065   -24.66355371 -1050.27363925
    14.45456523   -94.73026702   253.91014018   174.59792634
   144.49889341    11.72635993  -584.73331986   822.65987381
   611.96774428   568.72656448   338.46566497  -162.37509376
  -639.39791889 -1540.04683489  1548.42450016  1204.32012526
  1125.65860112   760.42618617   459.92459715  -151.51946186
 -1105.93621398 -3841.29833402]


In [22]:
#np.asarray(y_train)

In [23]:
#np.asarray(X_train)

In [24]:
import statsmodels.api as sm
X_train_sm = sm.add_constant(X_train)

In [25]:
#X_train_sm.head()

In [26]:
sm_OLS = sm.OLS(y_train, X_train_sm.astype(float)).fit()
sm_OLS.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.920
Model:                            OLS   Adj. R-squared:                  0.920
Method:                 Least Squares   F-statistic:                 2.013e+04
Date:                Fri, 11 Nov 2022   Prob (F-statistic):               0.00
Time:                        00:28:04   Log-Likelihood:            -3.4194e+05
No. Observations:               40451   AIC:                         6.839e+05
Df Residuals:                   40427   BIC:                         6.841e+05
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          3944.6634    313.682     12.575      0.000    3329.839    4559.488
carat           1.14e+04     59.512    191.618      0.000    1.13e+04    1.15e+04
depth           -61.9728      5.256    -11.792      0.000     -72.274     -51.672
table           -24.6636      3.360     -7.340      0.000     -31.250     -18.077
x             -1050.2736     37.566    -27.958      0.000   -1123.904    -976.643
y                14.4546     19.680      0.734      0.463     -24.119      53.028
z               -94.7303     38.703     -2.448      0.014    -170.589     -18.871
cut_Ideal      1042.8428     57.275     18.207      0.000     930.582    1155.104
cut_Premium     963.5306     61.819     15.586      0.000     842.364    1084.697
cut_Very Good   933.4316     61.968     15.063      0.000     811.972    1054.891
cut_Good        800.6590     66.688     12.006      0.000     669.948     931.370
cut_Fair        204.1994     77.187      2.646      0.008      52.911     355.488
color_D        1386.1832     47.056     29.458      0.000    1293.952    1478.415
color_E        1175.4911     46.478     25.291      0.000    1084.393    1266.589
color_F        1132.2499     46.642     24.275      0.000    1040.830    1223.670
color_G         901.9890     46.616     19.349      0.000     810.621     993.357
color_H         401.1482     47.051      8.526      0.000     308.928     493.369
color_I         -75.8746     47.847     -1.586      0.113    -169.657      17.908
color_J        -976.5235     50.154    -19.470      0.000   -1074.827    -878.220
clarity_IF     2041.5074     47.156     43.293      0.000    1949.081    2133.934
clarity_VVS1   1697.4031     43.477     39.041      0.000    1612.186    1782.620
clarity_VVS2   1618.7415     42.724     37.889      0.000    1535.002    1702.481
clarity_VS1    1253.5091     41.939     29.889      0.000    1171.307    1335.711
clarity_VS2     953.0075     41.764     22.819      0.000     871.149    1034.866
clarity_SI1     341.5635     42.399      8.056      0.000     258.461     424.666
clarity_SI2    -612.8533     42.749    -14.336      0.000    -696.643    -529.064
clarity_I1    -3348.2154     60.056    -55.752      0.000   -3465.926   -3230.505
==============================================================================
Omnibus:                    10780.214   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           528981.577
Skew:                           0.489   Prob(JB):                         0.00
Kurtosis:                      20.689   Cond. No.                     6.18e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 7.62e-28. This mig

In [27]:
OLS_params = sm_OLS.params
OLS_params

const             3944.663404
carat            11403.544293
depth              -61.972801
table              -24.663554
x                -1050.273639
y                   14.454565
z                  -94.730267
cut_Ideal         1042.842821
cut_Premium        963.530607
cut_Very Good      933.431574
cut_Good           800.659041
cut_Fair           204.199361
color_D           1386.183217
color_E           1175.491088
color_F           1132.249908
color_G            901.989008
color_H            401.148250
color_I            -75.874575
color_J           -976.523491
clarity_IF        2041.507426
clarity_VVS1      1697.403051
clarity_VVS2      1618.741527
clarity_VS1       1253.509112
clarity_VS2        953.007523
clarity_SI1        341.563464
clarity_SI2       -612.853288
clarity_I1       -3348.215408
dtype: float64

#### Technical point of model implementation comparison: The two model implementations return equivalent results and interpretations (sklearn.linear_model LinearRegression vs. statsmodels.api OLS)
The model used here is ordinary least-squares multivariate linear regression. The sklearn.linear_model LinearRegression and the statsmodels.api OLS are two different implementations of the same model.

Notice that the first six coefficients (carats to z) are the same between the models. However, the intercept and dummy variables have different coefficients. The training data R-squared values (0.920) are the same.

Because the dummy variables are either 1 or 0, they adjust the best-fit line up or down by the amount of the coefficient. The intercept also adjusts the best-fit line up or down by the amount of the intercept coefficient. Take the sum of the low values for cut, color, and clarity for the two models. The difference in those sums is exactly the difference between the intercepts of the model implementations. See cell below for validation.

The other important consideration in comparing the model implementations is to ensure that the relative differences between dummy variable coefficients is the same. See two cells below for validation of relative coefficient differences for `cut`. Similar validations can be done for `color` and `clarity` (not shown).

In [28]:
# Account for difference in intercept coefficients between model implementations.
# Notice that the difference in the sum of the low coefficients (1845.5) is balanced out by
# the difference in the intercept coefficients (-1845.5).
row_names = ['cut_Fair', 'color_J', 'clarity_I1', 'sum_lows', 'intercept']

lm_sum_lows = lm.coef_[10] + lm.coef_[17] + lm.coef_[25]
lm_lows = [lm.coef_[10], lm.coef_[17], lm.coef_[25], lm_sum_lows, lm.intercept_]

OLS_sum_lows = OLS_params.cut_Fair + OLS_params.color_J + OLS_params.clarity_I1
OLS_lows = [OLS_params.cut_Fair, OLS_params.color_J, OLS_params.clarity_I1, OLS_sum_lows, OLS_params.const]

OLS_lm_diff = [x - y for x, y in zip(OLS_lows, lm_lows)]

mod_comp_df = pd.DataFrame({'low names': row_names, 
                            'lm low coef': lm_lows,
                            'OLS low coef': OLS_lows,
                            'OLS - lm': OLS_lm_diff
                           })
mod_comp_df

low names  lm low coef  OLS low coef     OLS - lm
0    cut_Fair  -584.733320    204.199361   788.932681
1     color_J -1540.046835   -976.523491   563.523343
2  clarity_I1 -3841.298334  -3348.215408   493.082926
3    sum_lows -5966.078489  -4120.539539  1845.538950
4   intercept  5790.202354   3944.663404 -1845.538950

In [29]:
# Verify that relative dummy coefficient differences are the same between model implementations.
# Normalize the coefficients so that the smallest coefficient is 0 for ease of comparison.

lm_cut_norm = lm.coef_[6:11] - lm.coef_[10]
OLS_cut_norm = OLS_params[7:12] - OLS_params.cut_Fair
OLS_cut_norm.index
pd.DataFrame({'coef name': list(OLS_cut_norm.index),
              'lm coef norm': lm_cut_norm,
              'OLS coef norm': OLS_cut_norm
             })


coef name  lm coef norm  OLS coef norm
cut_Ideal          cut_Ideal    838.643460     838.643460
cut_Premium      cut_Premium    759.331246     759.331246
cut_Very Good  cut_Very Good    729.232213     729.232213
cut_Good            cut_Good    596.459680     596.459680
cut_Fair            cut_Fair      0.000000       0.000000